Expert series articles (#167)
heroiclabs/nakama-docs: docs/expert-tcp-udp.md

![image alt text](image_0.png)

If some messages must* *be received while others are merely nice to have, then the game’s networking code must support a gradient of reliability.

## Reliability is a gradient

Transport layer protocols partly map to the gradient by setting distinct expectations for message delivery.

UDP offers few reliability guarantees. It’s a more trusting protocol that expects that a message will be delivered. If it’s not delivered, the protocol won’t hear or do anything about it. It also doesn’t make any assumptions about when the message is delivered. UDP makes no effort to ensure the order of delivery. The first message you sent may not be the first received. When it comes to reliability, UDP is relaxed.

TCP offers robust reliability guarantees. It goes to some effort to confirm that a connection between the sender and recipient is established, that the recipient is prepared to receive messages, and, after the messages are sent, that they are delivered completely and in order. When it comes to reliability, TCP is a strict micromanager.

> ![image alt text](image_1.png)

Author: ```suggestion
![UDP vs TCP for reliability and overhead](nakama-tcp-vs-udp.png)
```

### A Tale of Four Messages

Under good network conditions, it’s actually quite hard to tell UDP and TCP apart in terms of reliability. But if conditions are poor, it’s easier to see how reliability tradeoffs emerge. To illustrate, suppose your game needs to send four messages sequentially (but let’s ignore their contents and importance, for now).

In good conditions, messages arrive intact, promptly, in-order, and just once. In poorer conditions, none of these assumptions hold. How do the different protocols cope?

With UDP, send the four messages, M1, M2, M3, and M4, each one after the other. What happens to each?

> ![image alt text](image_2.png)

Author: ```suggestion
![Visualization of operation described below.](nakama-udp.png)
```

* M1 gets lost along the way. It never arrives.

* M2 eventually arrives, but the routing is so convoluted, it’s last to arrive.

* M3 arrives first!

* M4 arrives, but a router along the way spontaneously duplicated the message in transit, so it arrives *twice*, once in second place and once in third.

As far as UDP is concerned, this is how it’s supposed to happen. It’s "best effort" but at least it’s low overhead. The protocol doesn’t require the receiver to notify the sender of any problems. If anything is to be done about this network hiccup—to make sense of out-of-order delivery, repeated delivery, or the lost message the receiver didn’t even know about—it has to be done at the application layer, in your game.

Let’s rewind things and play this scenario out again over TCP. TCP is a more complex protocol and we’ll have to tell a more complex story. But let’s send those messages again, assuming that we’ve already completed the three-way handshake and we’re in the ESTABLISHED state. What happens to each?

> ![image alt text](image_3.png)

Author: ```suggestion
![Visualization of operations described below.](nakama-tcp.png)
```

## Optionally reliable

Your game, as a networked application, can choose how to use the network.

In general, your game is not fixed to a single protocol: you can use UDP when you don’t require the reliability guarantees of TCP and you can switch to TCP when you do. Under good network conditions (low latency and packet loss), this won’t matter much. But if things get shaky, then you’ll avoid some misadventures associated with lost messages.

For points on the gradient between optional and required, there’s a strategy called *optionally reliable UDP*. This is when you build additional guarantees in your application, rather than relying on the protocol. In outline, this is sending UDP packets that require some response from the recipient; the message is re-sent until the expected response is received.

But whether you choose UDP or TCP or optionally reliable UDP comes down to the *content* of your messages and what they mean to your game and players.

## Start with your payloads 

The content of your messages matters more than the vehicle they travel in. Start by focusing on your payloads, then figure out how to get them there.

> Before you start sending messages, make sure you**_ _****know what you need to send over the wire** (and what you don’t). For example, in an action game, your player’s movements will probably be sent to the server many times per second, while your player’s inventory status only needs to be sent when it changes. Smaller data structures are less likely to pay fragmentation penalties. The fastest packet is the one you never send.

Reviewer: I think an asterisk typo here unless I'm missing some special sequence.

Author: Almost definitely an error, it seems the Google Docs -> Markdown generator I used has some interesting asterisk quirks. Thanks for catching all of these @lugehorsam!

Author: ```suggestion
Before you start sending messages, make sure you **know what you need to send over the wire** (and what you don’t). For example, in an action game, your player’s movements will probably be sent to the server many times per second, while your player’s inventory status only needs to be sent when it changes. Smaller data structures are less likely to pay fragmentation penalties. The fastest packet is the one you never send.
```

**Know your player’s context**. Your netcode should reflect when and how your player is interacting with the game. This starts with the player’s physical environment, such as playing on a desktop computer with a high-quality wired connection versus playing from a mobile device on a spotty mobile network. It also includes the game context; the game state may make certain kinds of messages more likely than others.

**Know what messages, if lost, would annoy the player** (or worse), what they’d never notice, and what’s in between. Each message’s place on the reliability gradient is key. If you know that prediction, interpolation, or other strategies can gracefully hide the occasional lost message, then that gives you more options for sending that message.

## Stand on shoulders of giants

If you’re thinking that this only begins to dig into thinking about net code, then you’re right. As you build out your games netcode, you’ll need to learn more about different aspects of prediction, interpolation, serialization, and beyond.

> Read Ruoyu Sun’s *[Game Networking Demystifie*d](https://ruoyusun.com/2019/03/28/game-networking-1.html), which takes a closer look at game states and client inputs and how they relate to cheating, server authority, and game genre.

Reviewer: asterisk typo here

Author: ```suggestion
Read Ruoyu Sun’s [*Game Networking Demystified*](https://ruoyusun.com/2019/03/28/game-networking-1.html), which takes a closer look at game states and client inputs and how they relate to cheating, server authority, and game genre.
```

> Another great read is *[Peeking into VALORANT's Netcod*e](https://technology.riotgames.com/news/peeking-valorants-netcode) from Riot Games, which goes in-depth on how product design goals can lead to specific decisions in netcode design.

Reviewer: asterisk typo here

Author: ```suggestion
Another great read is *[Peeking into VALORANT's Netcode](https://technology.riotgames.com/news/peeking-valorants-netcode)* from Riot Games, which goes in-depth on how product design goals can lead to specific decisions in netcode design.
```
